src/math_functions.cpp

```diff
@@ -0,0 +1,818 @@
+
+// TODO: cmath::M_PI has 3 more digits precision than the Fortran constant we
+// use so for now we will reuse the Fortran constant until we are OK with
+// modifying test results
+extern "C" const double PI {3.1415926535898};
+
+extern "C" const double SQRT_PI {std::sqrt(PI)};
+
+//==============================================================================
+// NORMAL_PERCENTILE calculates the percentile of the standard normal
+// distribution with a specified probability level
+//==============================================================================
+
+double __attribute__ ((const)) normal_percentile_c(const double p) {
+  double z;
+  double q;
+  double r;
+  const double p_low = 0.02425;
+  const double a[6] = {-3.969683028665376e1, 2.209460984245205e2,
+                       -2.759285104469687e2, 1.383577518672690e2,
+                       -3.066479806614716e1, 2.506628277459239e0};
+  const double b[5] = {-5.447609879822406e1, 1.615858368580409e2,
+                       -1.556989798598866e2, 6.680131188771972e1,
+                       -1.328068155288572e1};
+  const double c[6] = {-7.784894002430293e-3, -3.223964580411365e-1,
+                       -2.400758277161838, -2.549732539343734,
+                       4.374664141464968, 2.938163982698783};
+  const double d[4] = {7.784695709041462e-3, 3.224671290700398e-1,
+                       2.445134137142996, 3.754408661907416};
+
+  // The rational approximation used here is from an unpublished work at
+  // http://home.online.no/~pjacklam/notes/invnorm/
+
+  if (p < p_low) {
+    // Rational approximation for lower region.
+
+    q = std::sqrt(-2.0 * std::log(p));
+    z = (((((c[0]*q + c[1])*q + c[2])*q + c[3])*q + c[4])*q + c[5]) /
+          ((((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0);
+
+  } else if (p <= 1.0 - p_low) {
+    // Rational approximation for central region
+
+    q = p - 0.5;
+    r = q * q;
+    z = (((((a[0]*r + a[1])*r + a[2])*r + a[3])*r + a[4])*r + a[5])*q /
+         (((((b[0]*r + b[1])*r + b[2])*r + b[3])*r + b[4])*r + 1.0);
+
+  } else {
+    // Rational approximation for upper region
+
+    q = std::sqrt(-2.0*std::log(1.0 - p));
+    z = -(((((c[0]*q + c[1])*q + c[2])*q + c[3])*q + c[4])*q + c[5]) /
+          ((((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0);
+  }
+
+  // Refinement based on Newton's method
+
+  z = z - (0.5 * std::erfc(-z / std::sqrt(2.0)) - p) * std::sqrt(2.0 * PI) *
+       std::exp(0.5 * z * z);
+
+  return z;
+
+}
+
+//==============================================================================
+// T_PERCENTILE calculates the percentile of the Student's t distribution with
+// a specified probability level and number of degrees of freedom
+//==============================================================================
+
+double __attribute__ ((const)) t_percentile_c(const double p, const int df){
+  double t;
+  double n;
+  double k;
+  double z;
+  double z2;
```

Comment:
Declare `n`, `k`, `z`, and `z2` when they are used, and do the same for the rest of the functions here.

Reply:
- [x]